pennylane/templates/subroutines/commuting_evolution.py

```diff
@@ -0,0 +1,152 @@
+         computed gradients will be incorrect in general.
+
+    Args:
+        hamiltonian (.Hamiltonian): The commuting Hamiltonian defining the time-evolution operator.
+           The Hamiltonian must be explicitly written
+           in terms of products of Pauli gates (:class:`~.PauliX`, :class:`~.PauliY`,
+           :class:`~.PauliZ`, and :class:`~.Identity`).
+        time (int or float): The time of evolution, namely the parameter :math:`t` in :math:`e^{- i H t}`.
+
+    Keyword args:
+        frequencies (list[int or float]): The unique positive differences between eigenvalues in
+            the Hamiltonian's spectrum. If frequencies are not given, cost function partial
+            derivative will be computed using the standard two-term shift rule applied to the
+            constituent Pauli words in the Hamiltonian individually.
+
+        shifts (list[int or float]): The parameter shifts to use in obtaining the
+            generalized parameter shift rules. If unspecified, equidistant shifts are used.
+
+    .. UsageDetails::
+
+        The template is used inside a qnode:
+
+        .. code-block:: python
+
+            import pennylane as qml
+
+            n_wires = 2
+            dev = qml.device('default.qubit', wires=n_wires)
+
+            coeffs = [1, -1]
+            obs = [qml.PauliX(0) @ qml.PauliY(1), qml.PauliY(0) @ qml.PauliX(1)]
+            hamiltonian = qml.Hamiltonian(coeffs, obs)
+            frequencies = [2,4]
+
+            @qml.qnode(dev)
+            def circuit(time):
+                qml.PauliX(0)
+                qml.CommutingEvolution(hamiltonian, time, frequencies)
+                return qml.expval(qml.PauliZ(0))
+
+        >>> circuit(1)
+        0.6536436208636115
+    """
+
+    num_params = 3
+    num_wires = AnyWires
+    par_domain = "R"
+    grad_method = None
+
+    def __init__(self, hamiltonian, time, frequencies=None, shifts=None, do_queue=True, id=None):
+        from pennylane.gradients.general_shift_rules import (
+            get_shift_rule,
+        )  # pylint: disable=import-outside-toplevel
+
+        if not isinstance(hamiltonian, qml.Hamiltonian):
+            raise TypeError(
+                "hamiltonian must be of type pennylane.Hamiltonian, got {}".format(
+                    type(hamiltonian).__name__
+                )
+            )
+
+        if frequencies is not None:
+            self.grad_recipe = (get_shift_rule(frequencies, shifts)[0],) + (None,) * len(
+                hamiltonian.data
+            )
+            self.grad_method = "A"
+
+        self.hamiltonian = hamiltonian
+        self.num_params = len(hamiltonian.data) + 1
+        self.frequencies = frequencies
+        self.shifts = shifts
+
+        super().__init__(time, *hamiltonian.data, wires=hamiltonian.wires, do_queue=do_queue, id=id)
+
+        if qml.math.requires_grad(hamiltonian.coeffs):
+            self.grad_method = None  # fallback to decomposition

```

Comment:
Relatively minor, but if you structure it like

```python
trainable_hamiltonian = qml.math.requires_grad(hamiltonian.coeffs)

if frequencies is not None and not trainable_hamiltonian:
    self.grad_recipe = ...
    self.grad_method = "A"
```

then you could avoid the redundant call to `get_shift_rule` if the Hamiltonian is trainable.